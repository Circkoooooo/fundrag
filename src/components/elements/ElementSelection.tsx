import styled from 'styled-components'
import elementBucket from './index'
import { ElementComponentType } from './type'

const ElementSelectionContainer = styled.div`
	width: 100%;
`
const Selection = styled.div`
	padding: 10px;
`

const SelectionName = styled.div`
	font-weight: bold;
	margin-bottom: 10px;
`

const SelectionMain = styled.div`
	display: grid;
	grid-template-columns: 1fr 1fr;
	grid-gap: 10px;
`

const ElementDisplayAround = styled.div`
	padding: 10px;
	background-color: #f0f0f0;
	border-radius: 4px;
	cursor: pointer;
	transition: all ease-in-out 0.15s;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
`

interface ElementSelectionProps {
	pickElement?(element: ElementComponentType): void
}

/**
 * @description 将所有可以选择的组件渲染出来
 * @author onecirckoooooo
 * @date 2023/01/26 15:11
 */
const ElementSelection: React.FC<ElementSelectionProps> = ({ pickElement }) => {
	const renderGroupElements = () => {
		return elementBucket
			.filter((item) => item.elements.length !== 0)
			.map((item) => {
				return (
					<Selection key={crypto.randomUUID()}>
						<SelectionName>{item.name}</SelectionName>
						<SelectionMain>
							{item.elements.map((Element) => {
								return (
									<ElementDisplayAround
										key={crypto.randomUUID()}
										onClick={() =>
											pickElement && pickElement(Element)
										}
									>
										{Element.componentName}
									</ElementDisplayAround>
								)
							})}
						</SelectionMain>
					</Selection>
				)
			})
	}

	return (
		<ElementSelectionContainer>
			{renderGroupElements()}
		</ElementSelectionContainer>
	)
}

export default ElementSelection
