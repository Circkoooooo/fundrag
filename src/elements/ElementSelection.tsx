import { useEffect, useState } from 'react'
import styled from 'styled-components'
import elementBucket from './index'
import { chunkOutput, ElementFunctionComponentType } from './type'

const ElementSelectionContainer = styled.div`
	width: 100%;
`

const Selection = styled.div`
	padding: 10px;
`

const SelectionName = styled.div`
	font-weight: bold;
	margin-bottom: 10px;
`

const SelectionMain = styled.div`
	display: grid;
	grid-template-columns: 1fr 1fr;
	grid-gap: 10px;
`

interface ElementSelectionProps {
	pickElement?(element: ElementFunctionComponentType): void
}

const ElementDisplayAround = styled.div`
	padding: 10px;
	background-color: #f0f0f0;
	border-radius: 4px;
	cursor: pointer;
	transition: all ease-in-out 0.15s;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
`

/**
 * @description 将所有可以选择的组件渲染出来
 * @author onecirckoooooo
 * @date 2023/01/26 15:11
 */
const ElementSelection: React.FC<ElementSelectionProps> = ({ pickElement }) => {
	const [needRenderElements, setNeedRenderElements] = useState<chunkOutput[]>(
		[]
	)

	useEffect(() => {
		elementBucket.forEach((bucketItem) => {
			bucketItem.key = crypto.randomUUID()
		})
		setNeedRenderElements(
			elementBucket
				.filter((item) => item.elements.length !== 0) //筛选出含有子元素的容器元素
				.map((item) => {
					//给每一项element绑定unikey
					item.elements.forEach((element) => {
						element.defaultAppendProps.elementKey =
							crypto.randomUUID()
					})
					return item
				})
		)
	}, [])

	const renderGroupElements = () => {
		return needRenderElements.map((item) => {
			return (
				<Selection key={item.key}>
					<SelectionName>{item.name}</SelectionName>
					<SelectionMain>
						{item.elements.map((Element) => {
							return (
								<ElementDisplayAround
									key={Element.defaultAppendProps.elementKey}
									onClick={() =>
										pickElement && pickElement(Element)
									}
								>
									{Element.defaultAppendProps.componentName}
								</ElementDisplayAround>
							)
						})}
					</SelectionMain>
				</Selection>
			)
		})
	}

	return (
		<ElementSelectionContainer>
			{renderGroupElements()}
		</ElementSelectionContainer>
	)
}

export default ElementSelection
