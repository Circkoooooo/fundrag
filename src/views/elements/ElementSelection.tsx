import React from 'react'
import styled from 'styled-components'
import { DragComponentPackage } from './index'
import { DragComponent } from '../../encapsulator'

const Selection = styled.div`
	padding: 10px;
`

const SelectionName = styled.div`
	font-weight: bold;
	margin-bottom: 10px;
`

const SelectionMain = styled.div`
	display: grid;
	grid-template-columns: 1fr 1fr;
	grid-gap: 10px;
`

const ElementDisplayAround = styled.div`
	padding: 10px;
	background-color: #f0f0f0;
	border-radius: 4px;
	cursor: pointer;
	transition: all ease-in-out 0.15s;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
`

interface ElementSelectionProps {
	onItemClick: (component: DragComponent) => void
	componentPackages: DragComponentPackage[]
}

/**
 * 将所有可以选择的组件渲染出来
 */
const ElementSelection: React.FC<ElementSelectionProps> = ({ onItemClick, componentPackages }) => {
	return (
		<>
			{componentPackages.map((componentPkg) => {
				return (
					<Selection key={componentPkg.packageName.toString()}>
						<SelectionName>{componentPkg.packageName}</SelectionName>
						<SelectionMain>
							{componentPkg.components.map((component) => {
								return (
									<ElementDisplayAround
										key={component.componentRenderConfig.componentName.toString()}
										onClick={() => {
											component && onItemClick(component)
										}}
									>
										{component.componentRenderConfig.componentName}
									</ElementDisplayAround>
								)
							})}
						</SelectionMain>
					</Selection>
				)
			})}
		</>
	)
}

export default ElementSelection
