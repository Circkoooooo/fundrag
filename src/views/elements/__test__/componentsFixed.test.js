import renderer from 'react-test-renderer'
import ElementSelection from '../ElementSelection'

describe('component fixed', () => {
	it('ElementSelection', () => {
		const component = renderer.create(<ElementSelection />)
		expect(component).toMatchSnapshot()
	})
})
