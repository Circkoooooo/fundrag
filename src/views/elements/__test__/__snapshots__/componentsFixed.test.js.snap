// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component fixed ElementSelection 1`] = `
Array [
  <div
    className="sc-bcXHqe iZPgNH"
  >
    <div
      className="sc-gswNZR kRGSUb"
    >
      常用组件
    </div>
    <div
      className="sc-dkrFOg kQmwrY"
    >
      <div
        className="sc-hLBbgP daeFSo"
        onClick={[Function]}
      >
        文本组件
      </div>
      <div
        className="sc-hLBbgP daeFSo"
        onClick={[Function]}
      >
        测试组件
      </div>
    </div>
  </div>,
  <div
    className="sc-bcXHqe iZPgNH"
  >
    <div
      className="sc-gswNZR kRGSUb"
    >
      布局组件
    </div>
    <div
      className="sc-dkrFOg kQmwrY"
    >
      <div
        className="sc-hLBbgP daeFSo"
        onClick={[Function]}
      >
        线性组件
      </div>
    </div>
  </div>,
]
`;
